src/realm/sync/client_syncsocket.hpp

```diff
@@ -0,0 +1,235 @@
+
+    std::string address;   // Host address
+    port_type port;        // Host port number
+    std::string path;      // Includes access token in query.
+    std::string protocols; // One or more websocket protocols, separated with ", "
+    bool is_ssl;           // true if SSL should be used
+
+    /// The remaining fields are just pass through values from SyncConfig. They may
+    /// not be provided if the SDK chooses not to support the related config options.
+    /// These may be necessary when using websocket libraries without low-level
+    /// control. These items may be removed or converted to test only parameters
+    /// in the future.
+    std::map<std::string, std::string> headers; // Only includes "custom" headers.
+    bool verify_servers_ssl_certificate;        // If true, verify server ssl certs
+    util::Optional<std::string> ssl_trust_certificate_path;
+    std::function<SyncConfig::SSLVerifyCallback> ssl_verify_callback;
+    util::Optional<SyncConfig::ProxyConfig> proxy; // Send traffic through a proxy
+};
+
+
+/// The WebSocket base class that is used by the SyncClient to send data over the
+/// WebSocket connection with the server. This is the class that is returned by
+/// SyncSocketInterface::connect() when the connection to an endpoint is requested.
+/// If an error occurs while establishing the connection, the error is presented
+/// to the WebSocketObserver provided when the WebSocket was created.
+struct WebSocketInterface {
+    /// The destructor must close the websocket connection when the WebSocket object
+    /// is destroyed
+    virtual ~WebSocketInterface() = default;
+
+    /// Write data asynchronously to the WebSocket connection. The handler function
+    /// will be called when the data has been sent successfully. The WebSocketOberver
+    /// provided when the WebSocket was created will be called if any errors occur
+    /// during the write operation.
+    virtual void async_write_binary(const char* data, size_t size, util::UniqueFunction<void()>&& handler) = 0;
+};
+
+
+/// WebSocket observer interface in the SyncClient that receives the websocket
+/// events during operation.
+class WebSocketObserver {
+public:
+    virtual ~WebSocketObserver() = default;
+
+    /// Called when the websocket is connected, i.e. after the handshake is done.
+    /// The Sync Client is not allowed to send messages on the socket before the
+    /// handshake is complete and no message_received callbacks will be called
+    /// before the handshake is done.
+    ///
+    /// @param protocol The negotiated subprotocol value returned by the server
+    virtual void websocket_handshake_completion_handler(const std::string& protocol) = 0;
+
+    /// Called when an error occurs while establishing the WebSocket connection
+    /// to the server or during normal operations. No additional binary messages
+    /// will be processed after this function is called.
+    virtual void websocket_connect_error_handler() = 0;
```

Comment:
The websocket protocol doesn't mention any params - just fire an error event: https://websockets.spec.whatwg.org/#eventdef-websocket-error
It means that an error occurred and the connection is going to be closed. The actual reason is sent with the close event.

Reply:
Got it, so what should the client do when this event fires?

Reply:
I was thinking it would save a flag that an error occurred and not process any more incoming/outgoing message transfers and wait for the socket to be closed. It seems the close "event" should come immediately afterwards, but there may be a close handshake that needs to complete before the websocket is actually closed.